sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroSource.java

```diff
@@ -482,51 +486,56 @@ public AvroReader(AvroSource<T> source) {
       return (AvroSource<T>) super.getCurrentSource();
     }
 
+    // Precondition: the stream is positioned after the sync marker in the current (about to be
+    // previous) block. currentBlockSize equals the size of the current (previous) block, or zero
+    // if this reader was just started.
+    //
+    // Postcondition: same as above, but for the current (formerly next) block.
     @Override
     public boolean readNextBlock() throws IOException {
-      // The next block in the file is after the first sync marker that can be read starting from
-      // the current offset. First, we seek past the next sync marker, if it exists. After a sync
-      // marker is the start of a block. A block begins with the number of records contained in
-      // the block, encoded as a long, followed by the size of the block in bytes, encoded as a
-      // long. The currentOffset after this method should be last byte after this block, and the
-      // currentBlockOffset should be the start of the sync marker before this block.
-
-      // Seek to the next sync marker, if one exists.
-      currentOffset += advancePastNextSyncMarker(stream, getCurrentSource().getSyncMarker());
-
-      // The offset of the current block includes its preceding sync marker.
-      currentBlockOffset = currentOffset - getCurrentSource().getSyncMarker().length;
-
-      // Read a small buffer to parse the block header.
-      // We cannot use a BinaryDecoder to do this directly from the stream because a BinaryDecoder
-      // internally buffers data and we only want to read as many bytes from the stream as the size
-      // of the header. Though BinaryDecoder#InputStream returns an input stream that is aware of
-      // its internal buffering, we would have to re-wrap this input stream to seek for the next
-      // block in the file.
-      int read = stream.read(readBuffer);
-      // We reached the last sync marker in the file.
-      if (read <= 0) {
+      // First, we mark the previous block as read by adding the now-previous block size to the
+      // current block offset.
+      synchronized (progressLock) {
+        currentBlockOffset += currentBlockSizeBytes;
```

Comment:
Can you comment on why it is not racy to not hold this lock? Specifically, isn't it possible that something checks on the state, and sees that the currentBlockOffset has advanced but the currentBlockSizeBytes hasn't been updated yet.

Alternativeley, could we do the pre-reading, then get the lock and update both current values?